packages/store/addon/-private/system/record-array-manager.js

```diff
@@ -8,347 +8,804 @@ import { get, set } from '@ember/object';
+      }
 
-function destroy(entry) {
-  entry.destroy();
-}
+      pendingForIdentifier.add(identifier);
 
-function remove(array, item) {
-  let index = array.indexOf(item);
+      let pending = this._pendingIdentifiers;
+      let models = (pending[modelName] = pending[modelName] || []);
+      if (models.push(identifier) !== 1) {
+        return;
+      }
 
-  if (index !== -1) {
-    array.splice(index, 1);
-    return true;
-  }
+      emberRun.schedule('actions', this, this._flush);
+    }
 
-  return false;
-}
+    willDestroy() {
+      Object.keys(this._liveRecordArrays).forEach(modelName => this._liveRecordArrays[modelName].destroy());
+      this._adapterPopulatedRecordArrays.forEach(entry => entry.destroy());
+      this.isDestroyed = true;
+    }
 
-function updateLiveRecordArray(array, internalModels) {
-  let modelsToAdd = [];
-  let modelsToRemove = [];
+    destroy() {
+      this.isDestroying = true;
+      emberRun.schedule('actions', this, this.willDestroy);
+    }
+  };
 
-  for (let i = 0; i < internalModels.length; i++) {
-    let internalModel = internalModels[i];
-    let isDeleted = internalModel.isHiddenFromRecordArrays();
-    let recordArrays = internalModel._recordArrays;
+  const remove = function(array, item) {
+    let index = array.indexOf(item);
 
-    if (!isDeleted && !internalModel.isEmpty()) {
-      if (!recordArrays.has(array)) {
-        modelsToAdd.push(internalModel);
-        recordArrays.add(array);
+    if (index !== -1) {
+      array.splice(index, 1);
+      return true;
+    }
+
+    return false;
+  };
+
+  const updateLiveRecordArray = function(store, recordArray, identifiers) {
+    let identifiersToAdd = [];
+    let identifiersToRemove = [];
+
+    for (let i = 0; i < identifiers.length; i++) {
+      let identifier = identifiers[i];
+      let shouldInclude = shouldIncludeInRecordArrays(store, identifier);
+      let recordArrays = recordArraysForIdentifier(identifier);
+
+      if (shouldInclude) {
+        if (!recordArrays.has(recordArray)) {
+          identifiersToAdd.push(identifier);
+          recordArrays.add(recordArray);
+        }
+      }
+
+      if (!shouldInclude) {
+        identifiersToRemove.push(identifier);
+        recordArrays.delete(recordArray);
       }
     }
 
-    if (isDeleted) {
-      modelsToRemove.push(internalModel);
-      recordArrays.delete(array);
+    if (identifiersToAdd.length > 0) {
+      pushIdentifiers(recordArray, identifiersToAdd, internalModelFactoryFor(store));
     }
-  }
+    if (identifiersToRemove.length > 0) {
+      removeIdentifiers(recordArray, identifiersToRemove, internalModelFactoryFor(store));
+    }
+  };
 
-  if (modelsToAdd.length > 0) {
-    array._pushInternalModels(modelsToAdd);
-  }
-  if (modelsToRemove.length > 0) {
-    array._removeInternalModels(modelsToRemove);
-  }
-}
+  const pushIdentifiers = function(recordArray, identifiers, cache) {
+    if (RECORD_ARRAY_MANAGER_LEGACY_COMPAT && !recordArray._pushIdentifiers) {
```

Comment:
As discussed with Igor -

r/RECORD_ARRAY_MANAGER_LEGACY_COMPAT/REMOVE_RECORD_ARRAY_MANAGER_LEGACY_COMPAT 

(or something else/shorter.  Taking recommendations for a name!)

The problem - with the current name, we would turn from `false` to `true` (when `RECORD_ARRAY_MANAGER_IDENTIFIERS` is turned on)  and then finally `false` to indicate this block of code can be removed at some point.  This doesn't follow how the other feature flags work. 

So the path forward - `RECORD_ARRAY_MANAGER_IDENTIFIERS` will be turned to `true`.  However, this flag will still be `false`, thus this block of code will still apply for some period of time.  When we turn the flag to `true`, we are indicating we want to remove this block of code.  This should be a safe operation b/c the flag in question is effectively a no-op without `RECORD_ARRAY_MANAGER_IDENTIFIERS` on.

This should look something like...

```
if (!REMOVE_RECORD_ARRAY_MANAGER_LEGACY_COMPAT && !recordArray._pushIdentifiers) {
```